pkg/workflow/workflow.go

```diff
@@ -0,0 +1,232 @@
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	k8sRuntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2/klogr"
+	diskv1beta1 "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/azuredisk/v1beta1"
+)
+
+type workflowKey struct{}
+
+type Workflow struct {
+	name           string
+	errSet         *sync.Map
+	startTime      time.Time
+	requestID      string
+	logger         Logger
+	parentWorkflow *Workflow
+	numWait        int32 // numWait should always be number of children + 1(itself)
+}
+
+type Option func(*Workflow)
+
+func WithDetails(details ...interface{}) Option {
+	return func(w *Workflow) {
+		w.logger = Logger{w.logger.WithValues(details...)}
+	}
+}
+
+func GetWorkflowFromObj(ctx context.Context, obj k8sRuntime.Object, details ...interface{}) (context.Context, Workflow) {
+	requestID := GetRequestID(obj)
+	startTimeStr := getAnnotationFromRuntimeObject(obj, consts.RequestStartimeKey)
+	var startTime time.Time
+	if startTimeStr != "" {
+		startTime, _ = time.Parse(consts.RequestTimeFormat, startTimeStr)
+	}
+	w := Workflow{
+		requestID: requestID,
+		startTime: startTime,
+		logger:    Logger{klogr.New().WithValues(append(GetObjectDetails(obj), consts.RequestIDKey, requestID)...)},
+	}
+	w.parentWorkflow = &w
+	return w.SaveToContext(ctx), w
+}
+
+func GetObjectDetails(obj k8sRuntime.Object) []interface{} {
+	switch target := obj.(type) {
+	case *diskv1beta1.AzVolume:
+		if target != nil {
+			return []interface{}{consts.VolumeNameLabel, target.Spec.VolumeName, consts.PvNameKey, target.Spec.PersistentVolume}
+		}
+	case *diskv1beta1.AzVolumeAttachment:
+		if target != nil {
+			return []interface{}{consts.VolumeNameLabel, target.Spec.VolumeName, consts.NodeNameLabel, target.Spec.NodeName, consts.RoleLabel, target.Spec.RequestedRole}
+		}
+	}
+	return []interface{}{}
+}
+
+// New starts a new child workflow if there is already a workflow saved to the context otherwise a new workflow
+func New(ctx context.Context, opts ...Option) (context.Context, Workflow) {
+	name := getCallerName()
+
+	w := emptyWorkflow()
+	w.name = name
+
+	if parentWorkflow, ok := GetWorkflowFromContext(ctx); ok {
+		w.requestID = parentWorkflow.requestID
+		w.parentWorkflow = &parentWorkflow
+		w.logger = parentWorkflow.Logger()
+		_ = atomic.AddInt32(&w.parentWorkflow.numWait, 1)
+	} else {
+		w.requestID = GenerateRequestID()
+		w.logger = Logger{klogr.New().WithValues(consts.RequestIDKey, w.requestID)}
+	}
+
+	for _, opt := range opts {
+		opt(&w)
+	}
+
+	// so workflow's numWait will always be number of children workflow + 1 (itself)
+	return w.SaveToContext(ctx), w
+}
+
+func emptyWorkflow() Workflow {
+	w := Workflow{
+		startTime: time.Now(),
+		logger:    Logger{klogr.New()},
+		numWait:   1,
+		errSet:    &sync.Map{},
+	}
+	w.parentWorkflow = &w
+	return w
+}
+
+func (w Workflow) Finish(errs ...error) {
+	if len(errs) > 0 {
+		for _, err := range errs {
+			w.errSet.Store(err, nil)
+		}
+	}
+	if w.numWait > 0 {
```

Comment:
Since the `numWait` count includes this instance, should the atomic decrement happen here?

We should defend against calling `Finish` too many times and `panic` is the count goes negative.